core/src/main/scala/org/bykn/bosatsu/Normalization.scala

```diff
@@ -0,0 +1,666 @@
+  case class App(fn: NormalExpression, arg: NormalExpression)
+  extends NormalExpression {
+    val maxLambdaVar = (fn.maxLambdaVar.toList ++ arg.maxLambdaVar.toList)
+      .reduceLeftOption(Math.max)
+  }
+  case class ExternalVar(pack: PackageName, defName: Identifier)
+  extends NormalExpression {
+    val maxLambdaVar = None
+  }
+  case class Match(arg: NormalExpression,
+    branches: NonEmptyList[(NormalPattern, NormalExpression)])
+  extends NormalExpression {
+    val maxLambdaVar =
+      (arg.maxLambdaVar.toList ++ branches.toList.flatMap(_._2.maxLambdaVar))
+        .reduceLeftOption(Math.max)
+  }
+  case class LambdaVar(index: Int) extends NormalExpression {
+    val maxLambdaVar = Some(index)
+  }
+  /*
+   * It is reasonable to ask how you can define a lambda without an identifier
+   * for its argument in its expression. This is a benefit of de bruijn indexing.
+   * When all lambdas have exactly one argument you identify the var by how many
+   * lambdas out you have to travel.
+   *
+   * eg \x -> \y -> [y,x] would have normalization Lambda(Lambda(Apply(Apply(List, LambdaVar(1)), LambdaVar(0))))
+   *
+   * ref: https://en.wikipedia.org/wiki/De_Bruijn_index
+   */
+  case class Lambda(expr: NormalExpression) extends NormalExpression {
+    val maxLambdaVar = expr.maxLambdaVar.map(_ - 1)
+  }
+  case class Struct(enum: Int, args: List[NormalExpression]) extends NormalExpression {
+    val maxLambdaVar = args.flatMap(_.maxLambdaVar).reduceLeftOption(Math.max)
+  }
+  case class Literal(lit: Lit) extends NormalExpression {
+    val maxLambdaVar = None
+  }
+  case class Recursion(lambda: NormalExpression) extends NormalExpression {
+    val maxLambdaVar = lambda.maxLambdaVar
+  }
+}
+
+sealed abstract class NormalPattern {}
+
+object NormalPattern {
+  case object WildCard extends NormalPattern
+  case class Literal(toLit: Lit) extends NormalPattern
+  case class Var(name: Int) extends NormalPattern
+  /**
+   * Patterns like foo @ Some(_)
+   * @ binds tighter than |, so use ( ) with groups you want to bind
+   */
+  case class Named(name: Int, pat: NormalPattern) extends NormalPattern
+  case class ListPat(parts: List[Either[Option[Int], NormalPattern]]) extends NormalPattern
+  case class PositionalStruct(name: Option[Int], params: List[NormalPattern]) extends NormalPattern
+  case class Union(head: NormalPattern, rest: NonEmptyList[NormalPattern]) extends NormalPattern
+}
+
+object Normalization {
+  case class NormalExpressionTag(ne: NormalExpression, children: Set[NormalExpression])
+  type NormalizedPM = PackageMap.Typed[(Declaration, Normalization.NormalExpressionTag)]
+  type NormalizedPac = Package.Typed[(Declaration, Normalization.NormalExpressionTag)]
+
+  type PatternEnv = Map[Int, NormalExpression]
+
+  sealed trait PatternMatch[+A]
+  case class Matches[A](env: A) extends PatternMatch[A]
+  case object NoMatch extends PatternMatch[Nothing]
+  case object NotProvable extends PatternMatch[Nothing]
```

Comment:
what does does this branch mean? Is it kind of ill-typed, or unknown in some way? It would seem that well-typed programs either match or not.

Reply:
This is the most common branch. Remember we're dealing with NormalExpressions and not actual values. This code is just trying to figure out if there's enough information in the NormalExpression to pick a branch (with just static analysis).

So if you are matching on anything other than a `Struct` expression then you'll like end up here. It's a little misleading in the tests because virtually everything reduces unless we're in a lambda or using an external